packages/datatrak-web-server/src/utils/processSurveyResponse.ts

```diff
@@ -0,0 +1,125 @@
+/*
+ * Tupaia
+ *  Copyright (c) 2017 - 2023 Beyond Essential Systems Pty Ltd
+ */
+import { getBrowserTimeZone } from '@tupaia/utils';
+import moment from 'moment';
+import {
+  DatatrakWebSubmitSurveyRequest as RequestT,
+  MeditrakSurveyResponseRequest,
+} from '@tupaia/types';
+import { buildUpsertEntity } from './buildUpsertEntity';
+
+export const isUpsertEntityQuestion = (config?: RequestT.EntityQuestionConfig) => {
+  if (!config?.entity) {
+    return false;
+  }
+  if (config.entity.createNew) {
+    return true;
+  }
+  return config.entity.fields && Object.keys(config.entity.fields).length > 0;
+};
+
+// Process the survey response data into the format expected by the endpoint
+export const processSurveyResponse = async (
+  surveyResponseData: RequestT.ReqBody,
+  getEntity: Function,
+) => {
+  const {
+    userId,
+    surveyId,
+    countryId,
+    questions = [],
+    answers = {},
+    startTime,
+  } = surveyResponseData;
+  const timezone = getBrowserTimeZone();
+  const timestamp = moment().toISOString();
+  // Fields to be used in the survey response
+  const surveyResponse = {
+    user_id: userId,
+    survey_id: surveyId,
+    start_time: startTime,
+    entity_id: countryId,
+    country_id: countryId,
+    end_time: timestamp,
+    data_time: timestamp,
+    timestamp,
+    timezone,
+    entities_upserted: [],
+    options_created: [],
+    answers: [],
+  } as MeditrakSurveyResponseRequest;
+  // Process answers and save the response in the database
+  const answersToSubmit = [] as Record<string, unknown>[];
+
+  for (const question of questions) {
+    const { questionId, questionType, id } = question;
```

Comment:
I really dislike that `Question.id` isn't an id in the `question` table, but `Question.questionId` is. That feels like it's just inviting mistakes down the line.

Reply:
Yes that's a good point. There's actually a componentId for the id but I realised when I reviewed this that the id and the questionId are the same field and that the question if should not be getting sent with the answer so I have removed reference to it completely!